Kudu.Services/Diagnostics/ProcessController.cs

```diff
@@ -222,6 +236,199 @@ public HttpResponseMessage MiniDump(int id, int dumpType = 0, string format = nu
+                    response.Content.Headers.ContentDisposition = new ContentDispositionHeaderValue("attachment");
+                    response.Content.Headers.ContentDisposition.FileName = profileFileName;
+                    return response;
+                }
+                else
+                {
+                    try
+                    {
+                        FileSystemHelpers.DeleteFile(profileFileFullPath);
+                    }
+                    catch
+                    {
+                    }
+
+                    return null;
+                }
+            }
+        }
+
+        private async Task<HttpResponseMessage> HandleProfilingProcessAsync(string processName, string arguments)
+        {
+            MemoryStream outputStream = null;
+            MemoryStream errorStream = null;
+            try
+            {
+                _tracer.Step("ProcessName:" + processName + "   arguments:" + arguments );
+                var exe = new Executable(processName, Path.GetDirectoryName(processName), TimeSpan.FromSeconds(ProcessExitTimeoutInSeconds));
+
+                outputStream = new MemoryStream();
+                errorStream = new MemoryStream();
+
+                _tracer.Step("Path:" + exe.Path + " working directory:" + exe.WorkingDirectory);
+
+                int exitCode = await exe.ExecuteAsync(_tracer, arguments, outputStream, errorStream);
+
+                string output = GetString(outputStream);
+                string error = GetString(errorStream);
+
+                _tracer.Step(output);
+
+                if (exitCode != 0)
+                {
+                    _tracer.TraceError(string.Format(CultureInfo.InvariantCulture, "Starting the {0} failed with the following error code '{1}'.", processName, exitCode));
+                    return Request.CreateErrorResponse(HttpStatusCode.InternalServerError, "Profiling process failed with the following error code: " + exitCode);
+                }
+                else if (!string.IsNullOrEmpty(error))
+                {
+                    _tracer.TraceError(error);
+                    return Request.CreateErrorResponse(HttpStatusCode.InternalServerError, "Profiling process failed with the following error: " + error);
+                }
+
+                _tracer.Step(output);
+                return Request.CreateResponse(HttpStatusCode.OK);
+            }
+            catch(Exception ex)
+            {
+                _tracer.TraceError(ex);
+                return Request.CreateErrorResponse(HttpStatusCode.InternalServerError, ex.Message);
+            }
+            finally
+            {
+                if(outputStream != null)
+                {
+                    outputStream.Dispose();
+                }
+
+                if (errorStream != null)
+                {
+                    errorStream.Dispose();
+                }
+            }
+        }
+
+        private static int GetNextProfilingSessionId()
```

Comment:
I think it would be good to move all the new Profiling stuff you've added to this controller into its own class, e.g. ProfileManager or something. That keeps this API controller clean and focused on serving UI, with all the internal details of managing profiles encapsulated in the other type. It would also make all this new logic more unit testable.